CenterForOpenScience/osf.io
# [Feature] Normalize OSF Verification Key [#OSF-6560]

## Update

- Fix merge conflicts
- Do not reveal user emails in email links
- Normalize `verification key`
    - forgot and reset password
    - claim contributor-ship
    - confirm registration or email
    - update tests
- Temporary Fix for OSF-6673
- Several minor improvements
- The following will be moved to a new feature ticket: https://openscience.atlassian.net/browse/OSF-6998.
    - Users can resend confirmation email for contributor claim
    - Bug fixing for OSF-6673

## Purpose

### Issue

When user forgets and resets password, verification key never gets deleted unless overwritten. In addition, validation only checks the key. Neither the user to whom it belongs nor whether it has expired is checked. Furthermore, verification key are used twice instead of "one-time" for several tasks.

### Fix

Add `User.verification_key_v2` which contains both the token string and an expiration time. Validation now checks username, token and expiration time. Renew the key once used.

### Further Fix

There are two other functionality "User Claim Contributor-ship" and "Send/Resend Registration Confirmation Email" that have similar issue. Although they have their own way of verification, both are fixed with the same idea.

## Changes

### Normalized Verification Key

```python
### defaults.py
# Expiration time for verification key
EXPIRATION_TIME_DICT = {
    'password': 30,         # 30 minutes for forgot and reset password
    'confirm': 24 * 60,     # 24 hours in minutes for confirm account and email
    'claim': 30 * 24 * 60    # 30 days in minutes for claim contributor-ship
}
```
```python
### core.py
# generate verification key
def generate_verification_key(verification_type=None):
    """
    Generate a one-time verification key with an optional expiration time.
    The type of the verification key determines the expiration time defined in `website.settings.EXPIRATION_TIME_DICT`.

    :param verification_type: None, verify, confirm or claim
    :return: a string or a dictionary
    """
    token = security.random_string(30)
    # v1 with only the token
    if not verification_type:
        return token
    # v2 with a token and the expiration time
    expires = dt.datetime.utcnow() + dt.timedelta(minutes=settings.EXPIRATION_TIME_DICT[verification_type])
    return {
        'token': token,
        'expires': expires,
    }
```
```python
### core.py/User
    verification_key = fields.StringField()

    verification_key_v2 = fields.DictionaryField(default=dict)
    # Format: {
    #   'token': <the verification token>
    #   'expires': <the verification expiration time>
    # }

    unclaimed_records = fields.DictionaryField(required=False)
    # Format: {
    #   <project_id>: {
    #       'name': <name that referrer provided>,
    #       'referrer_id': <user ID of referrer>,
    #       'token': <the verification token>,
    #       'expires': <the verification expiration time>,
    #       'email': <email the referrer provided or None>,
    #       'claimer_email': <email the claimer entered or None>,
    #       'last_sent': <timestamp of last email sent to referrer or None>
    #   }
    #   ...
    # }

    email_verifications = fields.DictionaryField(default=dict)
    # Format: {
    #   <token> : {
    #       'email': <email address>,
    #       'expiration': <datetime>,
    #       'confirmed': whether user is confirmed or not,
    #       'external_identity': user's external identity,
    #   }
    # }
```

### Other

- Resend confirmation will refresh the token.
- Update `get_or_create_user()` with verification key.
- Remove duplicated code `get_claim_token()` and `get_confirm_token()`.
- Remove "lingering" code fragments due to previous merge conflicts.
- Update related routes, makos, factories and tests

## Side effects

- Password reset link becomes one time only. No longer valid once clicked.

## Ticket

https://openscience.atlassian.net/browse/OSF-6560
https://openscience.atlassian.net/browse/SEC-33
